[FLINK-2793] [dashboard] Redirect to leader in non-standalone mode

uce opened the pull request:
- Redirects HTTP requests to the web fronted of the leading job manager if not associated with leading job manager
- Splits up the start up of the job manager and web frontend (see `JobManager#runJobManager`) to first start the web fronted (binding to the port), then the job manager actor, then the web fronted with the address of the associated job manager.
- The port of the web frontend to redirect is determined via a newly introduced `JobManagerMessage`
- The associated job manager is used to determine whether the leader notification addresses the associated job manager or another one.
- Adds web front end tests via HttpTestClient (moved the generated files to the src folder for this) 

This breaks the leader notification abstraction in the sense that the job manager location is not transparent anymore. The reason for this is that the execution graph structures are not serializable (a proper fix would address this instead of redirecting). I have not integrated this with the TestingCluster.

rmetzger commented:
+1 The change looks good to merge. (maybe we should add the JIRA id when merging)

uce commented:
I would squash all commits and add the JIRA. Have you tried it out?

I would like to hear @tillrohrmann's take on this "abstraction" breaking behaviour before merging it.

rmetzger commented:
No, I didn't try it.

tillrohrmann commented:
I already opened the PR. Will review it asap.

On Thu, Oct 1, 2015 at 4:38 PM, Robert Metzger <notifications@github.com>
wrote:

> No, I didn't try it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/1202#issuecomment-144747629>.
>

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -129,6 +129,13 @@ class JobManager(
+   * The port of the web monitor as configured. Make sure that it is actually configured before
+   * starting the JobManager.
+   */
+  val webMonitorPort : Integer = flinkConfiguration.getInteger(
```
Why not using Scala's `Int` type here?

tillrohrmann replied:
Maybe we can add a comment which says that this is gonna be kicked out once we have made the `ExecutionGraph` properly serializable or so...

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1761,11 +1803,8 @@ object JobManager {
-      actorSystem: ActorSystem)
-    : WebMonitor = {
+  def startWebRuntimeMonitor(config: Configuration, leaderRetrievalService: LeaderRetrievalService,
+    actorSystem: ActorSystem) : WebMonitor = {
```
Line breaks not very scalaesque

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1773,8 +1812,7 @@ object JobManager {
       val ctor: Constructor[_ <: WebMonitor] = clazz.getConstructor(classOf[Configuration],
-        classOf[LeaderRetrievalService],
-        classOf[ActorSystem])
+        classOf[LeaderRetrievalService], classOf[ActorSystem])
```
Line breaks not very scalaesque

uce replied:
I was going for Kafkaesque... :D Just kidding, will change it.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java:
```diff
@@ -70,28 +104,68 @@ public ActorGateway getArchiveGateway() {
-				ActorRef jobManager = AkkaUtils.getActorRef(
-						leaderAddress,
-						actorSystem,
+				ActorRef jobManager = AkkaUtils.getActorRef(leaderAddress, actorSystem,
```
This is a blocking call which will block the `NodeCache's` thread which calls the `NodeCacheListener`. This should be avoided by using `AkkaUtils.getActorRefFuture` and a `Promise`.

uce replied:
Good catch

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java:
```diff
@@ -70,28 +104,68 @@ public ActorGateway getArchiveGateway() {
+							.ask(JobManagerMessages.getRequestWebMonitorPort(), timeout);
+
+					JobManagerMessages.ResponseWebMonitorPort portResponse =
+							(JobManagerMessages.ResponseWebMonitorPort) Await.result(portFuture, timeout);
```
Again a blocking call which is bad. Better to use futures to circumvent this problem.

tillrohrmann commented:
The PR looks good. I think we cannot do much about it being a temporary workaround for the limitation that our `ExecutionGraph` is not serializable. Once we have the resources this should be addressed, though.

I've got some minor comments. The main point is the blocking call in the `notifyLeaderAddress` of the `JobManagerArchiverRetriever`. This should be fixed imo.

Moreover, I think I've found a race condition. The `RuntimeMonitorHandler` checks whether the current leader is its own or a remote one. If it's the local JM, then the respective RequestHandler is called. Within the `RequestHandler`, such as `CurrentJobsOverviewHandler` the `JobManagerArchiverRetriever` is called again. If now the leader has changed to be a remote JM, then we'll have a problem. This could be fixed as well.

tillrohrmann commented:
What you can actually do to resolve the issue with the `RequestHandler` is to statically assign them the local JM's address instead of providing a `JobManagerArchiveRetriever`. Since they are only called when the local JM is the leader, they don't have to look up the current leader address.

uce commented:
Yes, good point. I will do this now.

uce commented:
OK. I've addressed your comments in the last two commits. Thanks again!

- The associated local leader gateway is only resolved once and getGateway will only return this one (new leader notifications don't change this one and only affect the redirect address)
- The handlers don't block on the redirect address, but just read the current redirect address.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java:
```diff
@@ -47,51 +60,160 @@
+	 * <p>This has to match the URL retrieved by the leader retrieval service. In tests setups you
+	 * have to make sure to use the correct type of URLs.
+	 */
+	public void setJobManagerAkkaUrlAndRetrieveGateway(String jobManagerAkkaUrl) throws Exception {
```
Actually, I meant to get rid of this extra redirect logic here.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/RuntimeMonitorHandler.java:
```diff
@@ -44,27 +47,40 @@
 	@Override
 	protected void channelRead0(ChannelHandlerContext ctx, Routed routed) throws Exception {
+		// Redirect to leader if necessary
+		String redirectAddress = retriever.getRedirectAddress();
```
Get rid of this `getRedirectAddress` method here and simply ask for the current leader. Then compare the leader address with the local job manager address to decide whether you have to redirect or not.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java:
```diff
@@ -150,97 +148,97 @@ else if (flinkRoot != null) {
-			.GET("/joboverview", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, true, true)))
-			.GET("/joboverview/running", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, true, false)))
-			.GET("/joboverview/completed", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, false, true)))
+			.GET("/joboverview", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, true, true), retriever))
```
Start the `RequestHandler` with the JobManager address of the local JobManager. They will only be called from the `RuntimeMonitorHandler` if the local JM is the leader.

tillrohrmann commented:
I've added some more comments. Ping me if it's too superficial.

uce commented:
It's not superficial at all. It's cleaner, because the retriever is independent of the redirect logic then.

I've decided against it though in order to keep the handlers cleaner (there only two different kinds atm though, so it would be OK to add it) and because the associated job manager URL is only known after start() is called, so I would have to update all handlers lazily after that.

This is due to the start up ordering of
1. start web server (bind to port)
2. get web server port and start job manager with web server port configured
3. set job manager address at web server

This ordering allows both components to pick random ports.

I just didn't want to change too much code this at this point, because I genuinely hope that we will make the execution graph serialisable after the release and get back to the old solution. But I'm also happy to change it if you like. Just ping me.

rmetzger commented:
Are there any blockers to this PR. Otherwise, I'd like to have it merged rather soon.

tillrohrmann commented:
Have fixed a possible race condition in `JobManagerRetriever` and rebased on the latest master. If Travis gives green light, I'll merge it.

rmetzger commented:
Great ;)

tillrohrmann commented:
Oh I forgot. I still want to test it on Yarn.
